wrappers/rust/icicle-cuda-runtime/src/memory.rs

```diff
@@ -117,6 +134,66 @@ impl<T> HostSlice<T> {
         self.0
             .iter_mut()
     }
+
+    pub fn is_pinnable(&self) -> bool {
+        let pinnable = get_device_attribute(cudaDeviceAttr::cudaDevAttrHostRegisterSupported, 0).unwrap();
+        let lockable =
+            get_device_attribute(cudaDeviceAttr::cudaDevAttrPageableMemoryAccessUsesHostPageTables, 0).unwrap();
+
+        pinnable == 1 && lockable == 0
+    }
+
+    pub fn pin(&self, flags: CudaHostRegisterFlags) -> CudaResult<()> {
+        unsafe {
+            let ptr = self.as_ptr() as *mut c_void;
+            let flags_to_set = flags.bits();
+            cudaHostRegister(ptr, self.len(), flags_to_set as c_uint).wrap()
+        }
+    }
+
+    pub fn unpin(&self) -> CudaResult<()> {
+        unsafe {
+            let mut flags = 0;
+            let ptr = self.as_ptr() as *mut c_void;
+            cudaHostGetFlags(&mut flags, ptr).wrap()?;
+            cudaHostUnregister(ptr).wrap()
+        }
+    }
+
+    pub fn allocate_pinned(count: usize, flags: CudaHostAllocFlags) -> CudaResult<&'static mut Self> {
```

Comment:
why is the returned HostSlice with static lifetime?

Reply:
Since the HostSlice that is returned is created in the function and we are returning a reference to it, it needs a static lifetime.